### PR TITLE
fix(tests): fix the set_sentry_option test helper

### PR DESCRIPTION
**Pull request by corps**
Fixes flakiness introduced by the combination of `region_silo_test` decorators in https://github.com/getsentry/sentry/pull/43329 and the `set_sentry_option` test helper, which did not follow python's `ContextManager` protocol correctly.

Basically, any context manager's `enter` and `exit` must behave like a stack -- for a given thread's stack, the most recent `enter` should be `exited` first (FILO).  Unfortunately, `set_entry_option` would `enter` in the middle of a test, but exit at that end of the test.  This interacts with `region_silo_test` which actually invokes `override_settings` itself to manage the `SiloMode` state within tests -- the "push" of the silo mode, while proceeding the exit of the `set_sentry_option`, would also "pop" proceeding the exit of the `set_sentry_option` configuration.  Because django's `override_settings` implementation reaches into the `settings._wrapped` object to manage its stack state (later restoring it in precisely the same order as the exit calls), this would result in a corrupt, inconsistent stack.

This was "flaky" because it isn't a problem right away.  It causes future tests to start in the wrong silo mode.  Depending on entropy of which next tests followed and how much they asserted, they could be running hybrid cloud code in tests not annotated or prepared for hybrid cloud specific logic.

The solution is simply to force the decorator to be a context manager, which allows these fixtures and the `region_silo_test` decorators to cooperator.

In addition, to prevent future such bad states, resulting in flaky tests nad hard to debug conditions, we now validate that the `SiloMode` has been cleaned up at the start and end of every test.

### CODE REVIEW

**Review by mikejihbe — APPROVED** [review]

**Review by RyanSkonnord — APPROVED** [review]
Good stuff; thanks for the thorough explanation.

File: tests/sentry/sentry_metrics/limiters/test_cardinality_limiter.py [hunk 1]
```diff
@@ -65,125 +66,119 @@ def use_quotas(
 
 
 def test_reject_all(set_sentry_option):
-    set_sentry_option(
+    with set_sentry_option(
```
Comment (jjbayer) [inline tests/sentry/sentry_metrics/limiters/test_cardinality_limiter.py:69]: In a case like this where the arguments to `set_sentry_option` do not depend on any function parameters, could `set_sentry_option` also be a decorator?

**Review by jjbayer — APPROVED** [review]

File: src/sentry/utils/pytest/fixtures.py [hunk 2]
```diff
@@ -432,16 +432,14 @@ def set_sentry_option(request):
...
+            do stuff
     ```
     """
     from sentry.testutils.helpers.options import override_options
 
     def inner(key, value):
-        ctx_mgr = override_options({key: value})
-        ctx_mgr.__enter__()
-
-        request.addfinalizer(lambda: ctx_mgr.__exit__(None, None, None))
```
Comment (markstory) [inline src/sentry/utils/pytest/fixtures.py:444]: Now that you're not using `request` do you still need the pytest fixture dependency?

File: tests/conftest.py [hunk 3]
```diff
@@ -142,3 +144,20 @@ def setup_default_hybrid_cloud_stubs():
...
+
+@pytest.fixture(autouse=True)
+def validate_silo_mode():
+    # NOTE!  Hybrid cloud uses many mechanisms to simulate multiple different configurations of the application
+    # during tests.  It depends upon `override_settings` using the correct contextmanager behaviors and correct
+    # thread handling in acceptance tests.  If you hit one of these, it's possible either that cleanup logic has
+    # a bug, or you may be using a contextmanager incorrectly.  Let us know and we can help!
+    if SiloMode.get_current_mode() != SiloMode.MONOLITH:
+        raise Exception(
+            "Possible test league bug!  SiloMode was not reset to Monolith between tests.  Please read the comment."
```
Comment (markstory) [inline tests/conftest.py:157]:
```suggestion
            "Possible test leak bug!  SiloMode was not reset to Monolith between tests.  Please read the comment for validate_silo_mode() in tests/conftest.py"
```

File: tests/conftest.py [hunk 4]
```diff
@@ -142,3 +144,20 @@ def setup_default_hybrid_cloud_stubs():
...
+        )
+    yield
+    if SiloMode.get_current_mode() != SiloMode.MONOLITH:
+        raise Exception(
+            "Possible test league bug!  SiloMode was not reset to Monolith between tests.  Please read the comment."
```
Comment (markstory) [inline tests/conftest.py:162]:
```suggestion
            "Possible test leak bug!  SiloMode was not reset to Monolith between tests.  Please read the comment for validate_silo_mode() in tests/conftest.py."
```

File: tests/sentry/sentry_metrics/limiters/test_cardinality_limiter.py (hunk 1 above)
Comment (corps) [reply tests/sentry/sentry_metrics/limiters/test_cardinality_limiter.py:69]: Definitely.  But if I'm honest, I'm inclined to leave that without making deeper changes for the time being.  Given that the current system causes subtle test bugs, I want one known "good" commit to separate HC changes from this context test bug.

File: src/sentry/utils/pytest/fixtures.py (hunk 2 above)
Comment (corps) [reply src/sentry/utils/pytest/fixtures.py:444]: Addressed

**Pull request merged**